src/controllers/bookController.ts

```diff
@@ -0,0 +1,128 @@
+import { NextFunction, Request, Response } from "express";
+import Book, { IBook } from "../models/Book";
+import { StatusCodes } from "http-status-codes";
+import { BadRequestError, NotFoundError } from "../errors";
+import mongoose from "mongoose";
+
+// Interface for query parameters
+interface IGetBooksQuery {
+  search?: string;
+  categories?: string;
+  sort?: "a-z" | "z-a" | "latest" | "oldest";
+}
+
+// Get all books
+const getAllBooks = async (
+  req: Request<{}, {}, {}, IGetBooksQuery>,
+  res: Response,
+  next: NextFunction
+) => {
+  const { search, categories, sort } = req.query;
+  let query: any = {};
+
+  // Filter by categories
+  if (categories) {
+    const categoryArray = Array.isArray(categories)
+      ? categories
+      : categories.split(",");
+
+    query.categories = { $in: categoryArray };
+  }
+
+  // Search by title
+  if (search) {
+    query.title = { $regex: new RegExp(search, "i") }; // Case-insensitive search
+  }
+
+  try {
+    // Sort options
+    const sortOptions: any = {};
```

Comment:
These sort options do not need to be in the `try` block. We should leave only the critical pieces of code that can throw exceptions in there.

Reply:
fixed